### PR TITLE
functorch userbenchmark

### PR DESCRIPTION
**Pull request by zou3519**
functorch userbenchmark

This PR sets up the functorch userbenchmark and configures it to run 3
functorch models in the torchbenchmark suite that have not made it to
stable (e.g. torchbench v3) yet, so we can get regression testing on
them.

In addition, this PR adds some microbenchmarks for functorch into the userbenchmark.

Test Plan:

`python run_benchmark.py functorch` gives:

```
(/raid/rzou/pt/binaries4-env) [1] rzou@devfair0317:~/pytorch/benchmark (functorch_userbenchmark) $ cat .userbenchmark/func
torch/metrics-20221007144500.json
{
    "name": "functorch",
    "environ": {
        "git_version": "b472915d42b840469910e8fb243d40441a4b5a17"
    },
    "metrics": {
        "lennard_jones_cpu": 4.601219408214092,
        "lennard_jones_cuda": 6.164539773017168,
        "functorch_maml_omniglot_cpu": 12597.586147487164,
        "functorch_maml_omniglot_cuda": 318.9176507294178,
        "functorch_dp_cifar10_cuda": 83.94835330545902,
        "vmap_hessian_fc_cuda": 60.61760187149048,
        "SimpleMLP_vmap_cpu": 0.46687781251966953,
        "SimpleMLP_vmap_cuda": 0.36797603964805603,
        "SimpleMLP_ensemble_multi_cpu": 2.3521081916987896,
        "SimpleMLP_ensemble_multi_cuda": 0.502417491748929,
        "SimpleCNN_ensemble_multi_cuda": 5.7795668579638,
        "SimpleMLP_ensemble_single_cpu": 2.3946920223534107,
        "SimpleMLP_ensemble_single_cuda": 0.5035349484533072,
        "SimpleCNN_ensemble_single_cuda": 5.614033993333578,
        "SimpleMLP_persamplegrad_cpu": 8.035034853965044,
        "SimpleMLP_persamplegrad_cuda": 1.9205754343420267,
        "SimpleCNN_persamplegrad_cuda": 3.8312537409365177
    }
}
```

### CODE REVIEW

File: torchbenchmark/models/lennard_jones/__init__.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def __init__(self, test, device, jit=False, batch_size=None, extra_args=[]):
         )
         self.model = self.model.to(device)
 
-        r = torch.linspace(0.5, 2 * sigma, steps=self.batch_size, requires_grad=True)
+        r = torch.linspace(0.5, 2 * sigma, steps=self.batch_size)
```
Comment (zou3519) [inline torchbenchmark/models/lennard_jones/__init__.py:63]: This is a bugfix -- the inputs to the model should not require grad, the weights of the model do.

File: userbenchmark/functorch/__init__.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    metrics = run_benchmarks()
+    result = {
+        'name': BM_NAME,
+        'environ': {
+            'git_version': torch.version.git_version,
+        },
+        'metrics': metrics,
+    }
+    pprint.pprint(result)
+    dump_output(BM_NAME, result)
```
Comment (zou3519) [inline userbenchmark/functorch/__init__.py:31]: From reading the other user benchmarks, it looks like the userbenchmark CI expects us to dump a json here. Is that correct?

Also, once the json has been dumped, does userbenchmark send the information to scribe?

File: userbenchmark/functorch/__init__.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    metrics = {}
+
+    for case_ctor in benchmark_cases:
+        case = case_ctor()
+        runtime_ms = benchmark(case)
+        metrics[case.name()] = runtime_ms
+    return metrics
+
+
+def run(args):
```
Comment (zou3519) [inline userbenchmark/functorch/__init__.py:21]: I don't know what `args` should be, this file currently ignores it.

File: userbenchmark/functorch/__init__.py (hunk 2 above)
Comment (xuzhao9) [reply userbenchmark/functorch/__init__.py:31]: Correct, the CI runs a command like `pytorch run_benchmark.py functorch`, then expect the json file to exist in the path like `benchmark/.userbenchmark/functorch/metric-YYMMDDHHMMSS.json`, then it will analyze the json and upload the result to Amazon S3 and scribe.

File: userbenchmark/functorch/__init__.py (hunk 3 above)
Comment (xuzhao9) [reply userbenchmark/functorch/__init__.py:21]: This is part of the interface between the benchmark driver and userbenchmark, because some benchmarks (e.g. distributed) want to pass in arguments. It is totally fine that functorch doesn't need extra arguments to run, but we suggest to annotate the `args` parameter to `List[str]`.

**Comment by xuzhao9** [comment]
You can test the benchmark using the command:
```
python run_benchmark.py functorch
```

And post the output file `REPO_ROOT/.userbenchmark/functorch/metrics-YYMMDDHHMMSS.json` in the PR.

**Comment by zou3519** [comment]
Thanks for the feedback, @xuzhao9. I ran the test and pasted the output into the PR body (and commit message), as well as annotated the `run(args)` function.

I also added some more microbenchmarks instead of creating a separate PR for them. Please let me know if it would be helpful to split up the PRs instead.

**Review by xuzhao9 — APPROVED** [review]
LGTM!
